### PR TITLE
Detach lifetime of output buffer from DecoderState

### PR DESCRIPTION
**Pull request by chyyran**
### Pull Request Overview
This pull request introduces **no new APIs** and makes no modifications to existing ones. There are no new features added. It is prep work for future features as per https://github.com/gendx/lzma-rs/issues/72.

This PR does two things.

* Removes ownership of `output` from `DecoderState`.
  * This removes the restriction of `DecoderState` to be parameterized on the lifetime of the output buffer to enable more flexible usage.
* Reuses already allocated buffers where possible for a small performance increase and additional flexibility for `DecoderState`
  * No public APIs have changed. Observable behaviour should be identical for existing consumers.

### Testing Strategy

This pull request was tested by...

- [x] The existing test suite is sufficient and should all pass. There should be no observable changes in behaviour or public API.

### CODE REVIEW

File: src/decode/lzma.rs [hunk 1]
```diff
@@ -188,33 +170,9 @@ where
...
-// Initialize decoder with circular buffer
-pub fn new_circular_with_memlimit<W>(
-    output: W,
-    params: LzmaParams,
-    memlimit: usize,
-) -> error::Result<DecoderState<W, lzbuffer::LzCircularBuffer<W>>>
-where
-    W: io::Write,
-{
+pub fn new_circular(params: LzmaParams) -> error::Result<DecoderState> {
```
Comment (gendx) [inline src/decode/lzma.rs:173]: With the `output` extracted out, this becomes identical to `new_accum`. Can you merge the two into a single function called `new`?

File: src/decode/lzma.rs [hunk 2]
```diff
@@ -152,20 +145,9 @@ where
...
-    lc: u32,
-    lp: u32,
-    pb: u32,
-    unpacked_size: Option<u64>,
-) -> DecoderState<W, lzbuffer::LzAccumBuffer<W>>
-where
-    W: io::Write,
-{
+pub fn new_accum(lc: u32, lp: u32, pb: u32, unpacked_size: Option<u64>) -> DecoderState {
     DecoderState {
```
Comment (gendx) [inline src/decode/lzma.rs:149]: I think this was missed before, but can we add some assertions on `lc`, `lp` and `pb` to prevent invalid values? I.e. `lc < 8`, `lp < 4` and `pb < 4`?

File: src/decode/lzma.rs [hunk 3]
```diff
@@ -239,48 +197,53 @@ where
...
 }
 
-impl<W, LZB> DecoderState<W, LZB>
-where
-    W: io::Write,
-    LZB: lzbuffer::LzBuffer<W>,
-{
+impl DecoderState {
     pub fn reset_state(&mut self, lc: u32, lp: u32, pb: u32) {
+        if self.lc + self.lp == lc + lp {
```
Comment (gendx) [inline src/decode/lzma.rs:202]: Likewise, the values of `lc`, `lp` and `pb` should be validated here.

File: src/decode/lzma.rs [hunk 4]
```diff
@@ -239,48 +197,53 @@ where
...
+            // we can just reset here..
```
Comment (gendx) [inline src/decode/lzma.rs:203]: Nit: "We can just reset here."

File: src/decode/lzma.rs [hunk 5]
```diff
@@ -239,48 +197,53 @@ where
...
+            self.literal_probs.iter_mut().for_each(|v| v.fill(0x400))
+        } else {
+            // need to reallocate because of new size of lc+lp
```
Comment (gendx) [inline src/decode/lzma.rs:206]: Nit: "We need to reallocate because of the new size of `lc+lp`."

File: src/decode/lzma.rs [hunk 6]
```diff
@@ -239,48 +197,53 @@ where
...
         rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
     ) -> error::Result<()> {
-        self.process_mode(rangecoder, ProcessingMode::Finish)
+        self.process_mode(output, rangecoder, ProcessingMode::Finish)
     }
 
     #[cfg(feature = "stream")]
     pub fn process_stream<'a, R: io::BufRead>(
         &mut self,
+        output: &mut LZB,
```
Comment (gendx) [inline src/decode/lzma.rs:243]: `LZB` isn't defined in this function.

File: src/decode/rangecoder.rs [hunk 7]
```diff
@@ -180,6 +180,11 @@ impl BitTree {
     ) -> io::Result<u32> {
         rangecoder.parse_reverse_bit_tree(self.num_bits, self.probs.as_mut_slice(), 0, update)
     }
+
+    #[inline]
```
Comment (gendx) [inline src/decode/rangecoder.rs:184]: This shouldn't be necessary. If this really brings a performance improvement (assuming the compiler didn't already inline it by default), let's do that in a separate PR with proper before/after benchmarks.

File: src/decode/rangecoder.rs [hunk 8]
```diff
@@ -215,4 +220,13 @@ impl LenDecoder {
             Ok(self.high_coder.parse(rangecoder, update)? as usize + 16)
         }
     }
+
+    #[inline]
```
Comment (gendx) [inline src/decode/rangecoder.rs:224]: Likewise here.

File: src/decode/stream.rs [hunk 9]
```diff
@@ -155,24 +156,23 @@ where
...
             Ok(params) => {
-                let decoder = if let Some(memlimit) = options.memlimit {
-                    new_circular_with_memlimit(output, params, memlimit)
-                } else {
-                    new_circular(output, params)
-                }?;
-
+                let dict_size = params.dict_size;
+                let decoder = new_circular(params)?;
+                let output =LzCircularBuffer::from_stream_with_memlimit(output, dict_size as usize,
```
Comment (gendx) [inline src/decode/stream.rs:161]: Is `LzCircularBuffer::from_stream` still used after this change? If not, let's just remove it and rename `from_stream_with_memlimit` to `from_stream`.

**Comment by gendx** [comment]
I've approved the workflows to run. As there are some formatting errors, could you fix them as well and squash your changes into a single commit (force-pushed to this branch), to have a clean git history?

File: src/decode/lzma.rs [hunk 10]
```diff
@@ -239,48 +197,53 @@ where
...
-impl<W, LZB> DecoderState<W, LZB>
-where
-    W: io::Write,
-    LZB: lzbuffer::LzBuffer<W>,
-{
+impl DecoderState {
     pub fn reset_state(&mut self, lc: u32, lp: u32, pb: u32) {
+        if self.lc + self.lp == lc + lp {
+            // we can just reset here..
+            self.literal_probs.iter_mut().for_each(|v| v.fill(0x400))
```
Comment (gendx) [inline src/decode/lzma.rs:204]: This requires Rust 1.50+. Can you increase the MSRV as such in https://github.com/gendx/lzma-rs/blob/master/.github/workflows/tests.yml#L14 and https://github.com/gendx/lzma-rs/blob/master/README.md?

**Review by gendx — CHANGES_REQUESTED** [review]

File: src/decode/lzma.rs (hunk 1 above)
Comment (chyyran) [reply src/decode/lzma.rs]: I've renamed this to `new` and moved it to an associated function of `DecoderState`

File: src/decode/lzma.rs (hunk 3 above)
Comment (chyyran) [reply src/decode/lzma.rs:182]: Done:

https://github.com/gendx/lzma-rs/pull/73/files#diff-768757151baad4fb0605aaeee5210dc343977125d64711ebcf5a0e755b2f23c0R177

File: src/decode/lzma.rs (hunk 2 above)
Comment (chyyran) [reply src/decode/lzma.rs]: https://github.com/gendx/lzma-rs/pull/73/files#diff-768757151baad4fb0605aaeee5210dc343977125d64711ebcf5a0e755b2f23c0R150

File: src/decode/stream.rs (hunk 9 above)
Comment (chyyran) [reply src/decode/stream.rs]: Renamed `LzCircularBuffer::from_stream_with_memlimit ` to `from_stream`. Also renamed `LzAccumBuffer::from_stream_with_memlimit` to `from_stream` to mirror the `LzCircularBuffer` API. The only consumer of `LzAccumBuffer::from_stream` was `lzma2`, where I just passed `usize::MAX` as the default limit.

File: src/decode/lzma.rs [hunk 11]
```diff
@@ -151,136 +144,86 @@ where
...
-        is_rep_g1: [0x400; 12],
-        is_rep_g2: [0x400; 12],
-        is_rep_0long: [0x400; 192],
-        state: 0,
-        rep: [0; 4],
-        len_decoder: rangecoder::LenDecoder::new(),
-        rep_len_decoder: rangecoder::LenDecoder::new(),
+impl DecoderState {
+    // Initialize decoder with accumulating buffer
+    pub fn new(lc: u32, lp: u32, pb: u32, unpacked_size: Option<u64>) -> Self {
```
Comment (gendx) [inline src/decode/lzma.rs:149]: Rather than passing the parameters one by one to the function, could this API use `LzmaParams` struct? This would avoid potential mistakes in swapping `lc`/`lp`/`pb` between the call site and the function definition. Or at the very least define a new `LzmaProperties` struct for `(lc, lp, pb)`, and use that for all functions that take as input these 3 parameters together.

File: src/decode/lzma.rs [hunk 12]
```diff
@@ -151,136 +144,86 @@ where
...
-) -> error::Result<DecoderState<W, lzbuffer::LzCircularBuffer<W>>>
-where
-    W: io::Write,
-{
-    new_circular_with_memlimit(output, params, std::usize::MAX)
-}
+    pub fn reset_state(&mut self, lc: u32, lp: u32, pb: u32) {
+        assert!(lc < 8);
+        assert!(lp < 4);
+        assert!(pb < 4);
```
Comment (gendx) [inline src/decode/lzma.rs:180]: I now realized that the ranges in the documentation are inconsistent with their actual definition in `LzmaParams` above:

```rust
        let lc: u32 = pb % 9;
        pb /= 9;
        let lp: u32 = pb % 5;
        pb /= 5;
```

Can you update the assertions, and also the docs to e.g. `0..=8`?

Ideally, the boundary condition on these values is also something we'd want to test (either as part of #33, or as new test cases - and also comparing to other implementations i.e. #20).

**Review by gendx — CHANGES_REQUESTED** [review]

File: src/decode/lzma.rs (hunk 12 above)
Comment (chyyran) [reply src/decode/lzma.rs]: Fixed this and moved assertions to `LzmaProperties::validate`

File: src/decode/lzma.rs (hunk 11 above)
Comment (chyyran) [reply src/decode/lzma.rs]: For now I've defined an `LzmaProperties` struct for `lclppb`. There may be a cleaner way of exposing this but this can be put off for bikeshedding a public raw decoder API.

File: src/decode/lzma2.rs [hunk 13]
```diff
@@ -138,24 +147,24 @@ where
 
             lzma_info!("Properties {{ lc: {}, lp: {}, pb: {} }}", lc, lp, pb);
         } else {
-            lc = decoder.lc;
-            lp = decoder.lp;
-            pb = decoder.pb;
+            lc = decoder.lzma_props.lc;
+            lp = decoder.lzma_props.lp;
+            pb = decoder.lzma_props.pb;
```
Comment (gendx) [inline src/decode/lzma2.rs:152]: Nit: This block could directly return an `LzmaProperties` struct, rather than extracting the properties here to re-merge them just afterwards.

File: src/decode/lzma.rs [hunk 14]
```diff
@@ -38,14 +37,39 @@ enum ProcessingStatus {
     Finished,
 }
 
-pub struct LzmaParams {
+#[derive(Copy, Clone)]
+/// LZMA 'lclppb' decompression properties.
+pub struct LzmaProperties {
     // most lc significant bits of previous byte are part of the literal context
```
Comment (gendx) [inline src/decode/lzma.rs:43]: Nit: This explanation can probably be merged with the following line, i.e.:

```
/// The number of literal context bits.
///
/// The most `lc` significant bits of the previous byte are part of the literal context.
```

File: src/decode/lzma.rs [hunk 15]
```diff
@@ -38,14 +37,39 @@ enum ProcessingStatus {
...
-    lp: u32, // 0..4
+    /// The number of literal context bits.
+    ///
+    /// `lc` must not be greater than 8.
+    pub lc: u32, // 0..=8
+    /// The number of literal position bits.
+    ///
+    /// `lp` must not be greater than 4.
+    pub lp: u32, // 0..=4
     // context for literal/match is plaintext offset modulo 2^pb
```
Comment (gendx) [inline src/decode/lzma.rs:52]: Likewise:

```
/// The number of position bits.
///
/// The context for literal/match is the plaintext offset modulo `2^pb`.
```

File: src/decode/lzma.rs [hunk 16]
```diff
@@ -151,136 +163,76 @@ where
...
-        is_rep_g0: [0x400; 12],
-        is_rep_g1: [0x400; 12],
-        is_rep_g2: [0x400; 12],
-        is_rep_0long: [0x400; 192],
-        state: 0,
-        rep: [0; 4],
-        len_decoder: rangecoder::LenDecoder::new(),
-        rep_len_decoder: rangecoder::LenDecoder::new(),
+impl DecoderState {
+    // Initialize decoder with accumulating buffer
```
Comment (gendx) [inline src/decode/lzma.rs:167]: There is no accumulating buffer anymore here. IMO the function is self-explanatory, the comment can be removed.

**Review by gendx — CHANGES_REQUESTED** [review]

File: src/decode/lzma.rs (hunk 16 above)
Comment (chyyran) [reply src/decode/lzma.rs]: https://github.com/gendx/lzma-rs/pull/73/commits/5716a6bca5872c33ff806ccfd6a87b6ea90e31b9#diff-768757151baad4fb0605aaeee5210dc343977125d64711ebcf5a0e755b2f23c0R167

File: src/decode/lzma.rs (hunk 14 above)
Comment (chyyran) [reply src/decode/lzma.rs]: https://github.com/gendx/lzma-rs/pull/73/commits/5716a6bca5872c33ff806ccfd6a87b6ea90e31b9#diff-768757151baad4fb0605aaeee5210dc343977125d64711ebcf5a0e755b2f23c0R45

File: src/decode/lzma2.rs (hunk 13 above)
Comment (chyyran) [reply src/decode/lzma2.rs]: https://github.com/gendx/lzma-rs/pull/73/commits/5716a6bca5872c33ff806ccfd6a87b6ea90e31b9#diff-aec280556b32db2cd7e0d9b67090c0833582fa5e5dac272202f84adcfd64e887R119

Moved the in-flight lc/lp/pb values inside of the `reset_props` blocks as well.

File: src/decode/lzma2.rs [hunk 17]
```diff
@@ -103,15 +112,15 @@ where
...
     }
 
     if reset_state {
-        let lc: u32;
-        let lp: u32;
-        let mut pb: u32;
+        let new_props = if reset_props {
+            let lc: u32;
+            let lp: u32;
+            let mut pb: u32;
```
Comment (gendx) [inline src/decode/lzma2.rs:122]: Please check and fix the lints in this function as reported by Clippy: https://github.com/gendx/lzma-rs/runs/6790343456

**Review by gendx — CHANGES_REQUESTED** [review]

**Comment by gendx** [comment]
Overall this looks very good! There are just some lint failures left, but you can ignore the ones in `src/decode/rangecoder.rs` as they are unrelated to this change (probably some new lints enabled recently, I'll make a separate PR to fix them).

File: src/decode/lzma2.rs (hunk 17 above)
Comment (chyyran) [reply src/decode/lzma2.rs]: fixed: https://github.com/gendx/lzma-rs/commit/163b034954e2c0295a8e25c5567559fdccce2470#diff-aec280556b32db2cd7e0d9b67090c0833582fa5e5dac272202f84adcfd64e887R124
